### PR TITLE
Fix H_Abstraction TS search heuristics for CH3 + RH <=> CH4 + R

### LINKED ISSUE #638: bc_length: ZeroDivisionError: float division by zero
**Describe the bug**

Appears that the branch chain length is equating to zero

```bash
Traceback (most recent call last):
  File "/home/calvin.p/Code/ARC//ARC.py", line 69, in <module>
    main()
  File "/home/calvin.p/Code/ARC//ARC.py", line 65, in main
    arc_object.execute()
  File "/home/calvin.p/Code/ARC/arc/main.py", line 583, in execute
    fine_only=self.fine_only,
  File "/home/calvin.p/Code/ARC/arc/scheduler.py", line 486, in __init__
    self.schedule_jobs()
  File "/home/calvin.p/Code/ARC/arc/scheduler.py", line 575, in schedule_jobs
    self.spawn_post_opt_jobs(label=label, job_name=job_name)
  File "/home/calvin.p/Code/ARC/arc/scheduler.py", line 1459, in spawn_post_opt_jobs
    self.spawn_ts_jobs()
  File "/home/calvin.p/Code/ARC/arc/scheduler.py", line 1483, in spawn_ts_jobs
    tsg=tsg_index,
  File "/home/calvin.p/Code/ARC/arc/scheduler.py", line 832, in run_job
    job.execute()
  File "/home/calvin.p/Code/ARC/arc/job/adapter.py", line 317, in execute
    self.execute_incore()
  File "/home/calvin.p/Code/ARC/arc/job/adapters/ts/heuristics.py", line 280, in execute_incore
    dihedral_increment=self.dihedral_increment,
  File "/home/calvin.p/Code/ARC/arc/job/adapters/ts/heuristics.py", line 1012, in h_abstraction
    reactants_reversed=reactants_reversed,
  File "/home/calvin.p/Code/ARC/arc/job/adapters/ts/heuristics.py", line 469, in combine_coordinates_with_redundant_atoms
    reactants_reversed=reactants_reversed,
  File "/home/calvin.p/Code/ARC/arc/job/adapters/ts/heuristics.py", line 668, in get_modified_params_from_zmat_2
    reactants_reversed=reactants_reversed,
  File "/home/calvin.p/Code/ARC/arc/job/adapters/ts/heuristics.py", line 703, in get_new_zmat_2_map
    spc_from_zmat_2 = ARCSpecies(label='spc_from_zmat_2', xyz=zmat_2_mod)
  File "/home/calvin.p/Code/ARC/arc/species/species.py", line 434, in __init__
    self.process_xyz(xyz)
  File "/home/calvin.p/Code/ARC/arc/species/species.py", line 1589, in process_xyz
    xyzs.append(remove_dummies(check_xyz_dict(xyz)))
  File "/home/calvin.p/Code/ARC/arc/species/converter.py", line 695, in check_xyz_dict
    xyz_dict = zmat_to_xyz(zmat=xyz_dict, keep_dummy=False)
  File "/home/calvin.p/Code/ARC/arc/species/converter.py", line 845, in zmat_to_xyz
    coords, symbols = zmat_to_coords(zmat, keep_dummy=keep_dummy)
  File "/home/calvin.p/Code/ARC/arc/species/zmat.py", line 1021, in zmat_to_coords
    coords = _add_nth_atom_to_coords(zmat=zmat, 

coords=coords, i=i, coords_to_skip=coords_to_skip)
  File "/home/calvin.p/Code/ARC/arc/species/zmat.py", line 1098, in _add_nth_atom_to_coords
    ubc = [(coords[c_index][0] - coords[b_index][0]) / bc_length,
ZeroDivisionError: float division by zero
```

**How to reproduce**
Insert code to reproduce this behavior.

**Expected behavior**

**Additional context**

Here is my input file: [input.zip](https://github.com/ReactionMechanismGenerator/ARC/files/11185501/input.zip)

However, this has been occurring on multiple input files

Possibly related to #565

### PR DESCRIPTION
**Pull request by alongd**
In the case of the reaction `CH3 + RH <=> CH4 + R`, we get that the ZMat constructed for CH4 has parameters which rely on the abstracted H atom. This is an edge case, usually higher angles and dihedral angles are not defined w.r.t H atoms where possible, but here there's no choice. The issue reported by @naddeu was that removing the H atom from the ZMat resulted in illegal ZMat parameters (a parameter that refers twice to a certain atom).
This PR fixes this issue and adds a bunch of tests

fixes #638

### CODE REVIEW

**Comment by Laxzal** [comment]
@alongd Looks good to me 

However, it appears to be failing on the `zmat.remove_1st_atom` w.r.t zmat_5.

```
E       -           'A_3_0_1': 63.02451429659178,
E       ?                              ^ ^ ^^^ ^
E       
E       +           'A_3_0_1': 63.024513203726876,
```

Looks like the calculated and expected are slightly off. However, when I pull this branch down and run the test, it passes fine... So I am not sure what is going on?

**Review by Laxzal — APPROVED** [review]

**Review by calvinp0 — DISMISSED** [review]

File: arc/job/adapters/ts/heuristics_test.py [hunk 1]
```diff
@@ -901,6 +901,23 @@ def test_heuristics_for_h_abstraction(self):
...
+        self.assertEqual(rxn1.family.label, 'H_Abstraction')
+        heuristics_1 = HeuristicsAdapter(job_type='tsg',
+                                         reactions=[rxn1],
+                                         testing=True,
+                                         project='test',
+                                         project_directory=os.path.join(ARC_PATH, 'arc', 'testing', 'heuristics'),
+                                         dihedral_increment=120,
+                                         )
+        heuristics_1.execute_incore()
+        self.assertEqual(len(rxn1.ts_species.ts_guesses), 3)
```
Comment (kfir4444) [inline arc/job/adapters/ts/heuristics_test.py:919]: Why should be exactly 3 ts guesses for this reaction?

File: arc/species/vectors_test.py [hunk 2]
```diff
@@ -246,6 +232,19 @@ def test_calculate_dihedral_angle(self):
...
+        distance = vectors.calculate_param(coords=self.propene['coords'], atoms=[0, 1])
+        self.assertAlmostEqual(distance, 1.497630871004034)
+        angle = vectors.calculate_param(coords=self.propene['coords'], atoms=[0, 1, 2])
+        self.assertAlmostEqual(angle, 125.18344391469404)
+        dihedral = vectors.calculate_param(coords=self.propene['coords'], atoms=[0, 1, 2, 6])
+        self.assertAlmostEqual(dihedral, 180.0)
+        with self.assertRaises(ValueError):
+            vectors.calculate_param(coords=self.propene['coords'], atoms=[0])
+        with self.assertRaises(ValueError):
+            vectors.calculate_param(coords=self.propene['coords'], atoms=[0, 1, 2, 3, 4])
```
Comment (kfir4444) [inline arc/species/vectors_test.py:246]: Could you add a test for the exception? like 1 or 5 atoms?

**Review by kfir4444 — COMMENTED** [review]
Thanks @calvinp0, a minor comment and question

File: arc/species/zmat_test.py [hunk 3]
```diff
@@ -1692,24 +1694,30 @@ def test_remove_1st_atom(self):
...
                            'coords': ((None, None, None), ('R_1_0', None, None), ('R_2_0', 'A_2_0_1', None),
                                       ('R_3_0', 'A_3_0_1', 'D_3_0_1_2')),
                            'vars': {'R_1_0': 1.0921994253661749, 'R_2_0': 1.092199370793132, 'A_2_0_1': 109.47122048587586,
                                     'R_3_0': 1.0921994253661749, 'A_3_0_0': 109.47122150322166, 'D_3_0_0_2': 239.99999891956398,
                                     'A_3_0_1': 63.024513203726876, 'D_3_0_1_2': 287.0894016092067},
                            'map': {0: 0, 1: 1, 2: 2, 3: 3}}
-        self.assertEqual(zmat.remove_1st_atom(zmat_5), expected_zmat_5)
+        self.assertEqual(zmat_5['symbols'], expected_zmat_5['symbols'])
+        self.assertEqual(zmat_5['map'], expected_zmat_5['map'])
+        #self.assertAlmostEqual(zmat_5['vars'], expected_zmat_5['vars'])
```
Comment (alongd) [inline arc/species/zmat_test.py:1718]: is this commented out on purpose? `expected_zmat_5` above is not used

File: arc/species/vectors_test.py (hunk 2 above)
Comment (calvinp0) [reply arc/species/vectors_test.py:246]: Hey @kfir4444 I think the code you commented on fulfil your requirements. Unless I am misunderstanding?

**Comment by alongd** [comment]
@kfir4444 and @calvinp0, are we all good to merge this?

**Comment by calvinp0** [comment]
> @kfir4444 and @calvinp0, are we all good to merge this?

I am happy with it

**Review by calvinp0 — APPROVED** [review]

**Comment by kfir4444** [comment]
> @kfir4444 and @calvinp0, are we all good to merge this?

100%

**Comment by alongd** [comment]
Thanks!

**Pull request merged**